File: DCB.py
#
#---Kate---
#---Project-1--
#---sign--HELLOIAMKATE---
#---Owner--Kate---
#---MB--Kate---
#---1001---
#
import discord
from discord import app_commands
from discord.ext import commands
import os
from dotenv import load_dotenv
from typing import Union
from discord.ext.commands import CooldownMapping
from datetime import datetime, timedelta
from KA import keep_alive  # Add this import
import asyncio
import json
import time 

# Load environment variables
load_dotenv()
TOKEN = os.getenv("TOKEN_SM")
if not TOKEN:
    raise ValueError("❌ Bot token not found in .env file")

class DataStorage:
    def __init__(self, bot):
        self.bot = bot
        self.data_file = "bot_data.json"
        self.save_lock = asyncio.Lock()
        self.bot.stored_embeds = {}  # Add this line

    async def save_data(self):
        async with self.save_lock:
            data = {
                'sticky_messages': self.bot.sticky_messages,
                'sticky_cooldowns': self.bot.sticky_cooldowns,
                'guild_sticky_messages': getattr(self.bot, 'guild_sticky_messages', {}),
                'server_info': self.bot.server_info,
                'stored_embeds': self.bot.stored_embeds  # Add this line
            }
            try:
                with open(self.data_file, 'w') as f:
                    json.dump(data, f, indent=4)
            except Exception as e:
                print(f"Error saving data: {e}")

    async def load_data(self):
        try:
            with open(self.data_file, 'r') as f:
                data = json.load(f)
                self.bot.sticky_messages = data.get('sticky_messages', {})
                self.bot.sticky_cooldowns = data.get('sticky_cooldowns', {})
                self.bot.guild_sticky_messages = data.get('guild_sticky_messages', {})
                self.bot.server_info = data.get('server_info', {})
                self.bot.stored_embeds = data.get('stored_embeds', {})  # Add this line
        except FileNotFoundError:
            print("No existing data file found, starting fresh")
            self.bot.stored_embeds = {}  # Add this line
        except Exception as e:
            print(f"Error loading data: {e}")

    async def auto_save(self):
        while True:
            await asyncio.sleep(300)  # Save every 5 minutes
            await self.save_data()

class Bot(commands.Bot):
    def __init__(self):
        intents = discord.Intents.all()
        self.server_info = {}
        self.sticky_messages = {}
        self.sticky_cooldowns = {}
        self.sticky_last_sent = {}
        self.guild_sticky_messages = {}
        self.data_storage = DataStorage(self)
        
        # Initialize bot without prefix
        super().__init__(
            command_prefix=None,  # Remove prefix
            intents=intents,
            activity=discord.Game(name="/help | Server Manager"),
            status=discord.Status.online,
            help_command=None  # Disable default help command
        )
    
    async def reload_bot(self):
        """Reload the bot and sync commands"""
        try:
            await self.tree.sync()
            print("✅ Commands synced!")
            return True
        except Exception as e:
            print(f"❌ Error syncing commands: {e}")
            return False
    
    async def update_sticky_message(self, channel_id: int, message_id: int, content: Union[str, dict]):
        """Update sticky message in cache"""
        self.sticky_messages[channel_id] = {
            'message_id': message_id,
            'content': content,
            'is_embed': isinstance(content, dict)
        }

    async def setup_hook(self):
        """Called before the bot starts running"""
        # Load saved data
        await self.data_storage.load_data()
        
        # Load extensions
        for command_module in [
            'commands.moderation',
            'commands.roles',
            'commands.utility',
            'commands.help',
            'commands.embed_logic_command',
            'commands.sticky'
        ]:
            try:
                await self.load_extension(command_module)
                print(f"✅ Loaded {command_module}")
            except Exception as e:
                print(f"❌ Error loading {command_module}: {e}")
        
        # Start auto-save task
        self.auto_save_task = self.loop.create_task(self.data_storage.auto_save())
        
        await self.tree.sync()
        print("✅ Commands synced!")

    async def update_server_info(self, guild):
        """Update server information cache"""
        self.server_info[guild.id] = {
            'name': guild.name,
            'id': guild.id,
            'member_count': guild.member_count,
            'icon_url': guild.icon.url if guild.icon else None,
            'banner_url': guild.banner.url if guild.banner else None,
            'description': guild.description or "No description set",
            'owner': guild.owner,
            'created_at': guild.created_at,
            'roles': len(guild.roles),
            'channels': len(guild.channels),
            'boost_level': guild.premium_tier,
            'boost_count': guild.premium_subscription_count
        }

    async def on_command_error(self, ctx, error):
        """Handle traditional command errors"""
        if isinstance(error, commands.errors.CommandNotFound):
            await ctx.send("❌ Command not found!")
        elif isinstance(error, commands.errors.MissingPermissions):
            await ctx.send("❌ You don't have permission to use this command!")
        else:
            print(f"An error occurred: {str(error)}")
            await ctx.send("❌ An error occurred while executing the command!")

bot = Bot()

async def load_extensions():
    for command_module in [
        'commands.moderation',
        'commands.roles',
        'commands.utility',
        'commands.help',
        'commands.embed_logic_command',
        'commands.sticky'
    ]:
        try:
            await bot.load_extension(command_module)
            print(f"✅ Loaded {command_module}")
        except Exception as e:
            print(f"❌ Error loading {command_module}: {e}")

@bot.event
async def on_message(message):
    if message.author.bot:
        return
    
    channel_id = message.channel.id
    if channel_id in bot.sticky_messages:
        # Check cooldown
        now = datetime.utcnow()
        last_sent = bot.sticky_last_sent.get(channel_id)
        cooldown = bot.sticky_cooldowns.get(channel_id, 1)
        
        if last_sent and (now - last_sent).total_seconds() < 0.005:  # 5ms check
            return  # Still on cooldown
            
        sticky_data = bot.sticky_messages[channel_id]
        
        # Delete ALL previous sticky messages to prevent duplication
        async for old_message in message.channel.history(limit=50):  # Increased search limit
            if old_message.author == bot.user and (
                old_message.id == sticky_data.get('message_id') or
                (old_message.embeds and sticky_data.get('is_embed')) or
                (not old_message.embeds and not sticky_data.get('is_embed'))
            ):
                try:
                    await old_message.delete()
                except discord.NotFound:
                    pass
        
        # Wait briefly to ensure message order
        await asyncio.sleep(0.1)
        
        # Send new sticky message
        content = sticky_data['content']
        if isinstance(content, dict) and content.get('type') == 'stored_embed':
            # Use stored embed
            embed_data = content['original_data']
            embed = discord.Embed(
                title=embed_data['title'],
                description=embed_data['description'],
                color=discord.Color(embed_data['color'])
            )
            if embed_data['footer']:
                embed.set_footer(text=f"{embed_data['footer']} • 📌 Sticky message • {message.guild.name}")
            else:
                embed.set_footer(text=f"📌 Sticky message • {message.guild.name}")
            if embed_data['thumbnail']:
                embed.set_thumbnail(url=embed_data['thumbnail'])
            if embed_data['image']:
                embed.set_image(url=embed_data['image'])
            
            new_sticky = await message.channel.send(embed=embed)
        elif sticky_data.get('is_embed', False):
            content = sticky_data['content']
            embed = discord.Embed(
                title=content.get('title', 'Sticky Message'),
                description=content.get('description'),
                color=getattr(discord.Color, content.get('color', 'blue'))()
            )
            embed.set_footer(text=f"📌 Sticky message • {message.guild.name}")
            embed.timestamp = discord.utils.utcnow()
            new_sticky = await message.channel.send(embed=embed)
        else:
            new_sticky = await message.channel.send(sticky_data['content'])
        
        # Update tracking
        bot.sticky_last_sent[channel_id] = now
        await bot.update_sticky_message(
            channel_id,
            new_sticky.id,
            sticky_data['content']
        )

#
#---Kate---
#---Project-1--
#---sign--HELLOIAMKATE---
#---Owner--Kate---
#---MB--Kate---
#---1001---
#

if __name__ == "__main__":
    try:
        keep_alive()
        print("✅ Web server started!")
        print("✅ Starting bot with data persistence...")
        bot.run(TOKEN)
    except Exception as e:
        print(f"❌ Error starting bot: {e}")
        # Save data on shutdown
        if hasattr(bot, 'data_storage'):
            asyncio.run(bot.data_storage.save_data())

File: DCB.js
const { Client, GatewayIntentBits, Collection, EmbedBuilder, ActivityType } = require('discord.js');
const dotenv = require('dotenv');
const fs = require('fs').promises;
const path = require('path');
const { keepAlive } = require('./KA.js');

// Load environment variables
dotenv.config();
const TOKEN = process.env.TOKEN_SM;
if (!TOKEN) {
    throw new Error("❌ Bot token not found in .env file");
}

class DataStorage {
    constructor(bot) {
        this.bot = bot;
        this.dataFile = "bot_data.json";
        this.bot.storedEmbeds = {};
    }

    async saveData() {
        const data = {
            stickyMessages: this.bot.stickyMessages,
            stickyCooldowns: this.bot.stickyCooldowns,
            guildStickyMessages: this.bot.guildStickyMessages || {},
            serverInfo: this.bot.serverInfo,
            storedEmbeds: this.bot.storedEmbeds
        };

        try {
            await fs.writeFile(this.dataFile, JSON.stringify(data, null, 4));
        } catch (e) {
            console.error(`Error saving data: ${e}`);
        }
    }

    async loadData() {
        try {
            const data = JSON.parse(await fs.readFile(this.dataFile, 'utf8'));
            this.bot.stickyMessages = data.stickyMessages || {};
            this.bot.stickyCooldowns = data.stickyCooldowns || {};
            this.bot.guildStickyMessages = data.guildStickyMessages || {};
            this.bot.serverInfo = data.serverInfo || {};
            this.bot.storedEmbeds = data.storedEmbeds || {};
        } catch (e) {
            if (e.code === 'ENOENT') {
                console.log("No existing data file found, starting fresh");
                this.bot.storedEmbeds = {};
            } else {
                console.error(`Error loading data: ${e}`);
            }
        }
    }

    async autoSave() {
        setInterval(() => this.saveData(), 300000); // Save every 5 minutes
    }
}

class Bot extends Client {
    constructor() {
        super({
            intents: [
                GatewayIntentBits.Guilds,
                GatewayIntentBits.GuildMessages,
                GatewayIntentBits.MessageContent,
                GatewayIntentBits.GuildMembers
            ]
        });

        this.serverInfo = new Map();
        this.stickyMessages = new Map();
        this.stickyCooldowns = new Map();
        this.stickyLastSent = {};
        this.guildStickyMessages = {};
        this.dataStorage = new DataStorage(this);
        this.commands = new Collection();
    }

    async reloadBot() {
        try {
            await this.application?.commands.set([...this.commands.values()]);
            console.log("✅ Commands synced!");
            return true;
        } catch (e) {
            console.error(`❌ Error syncing commands: ${e}`);
            return false;
        }
    }

    async updateStickyMessage(channelId, messageId, content) {
        this.stickyMessages.set(channelId, {
            messageId,
            content,
            isEmbed: typeof content === 'object'
        });
    }

    async init() {
        await this.dataStorage.loadData();
        
        // Load command files directly from commands folder first
        try {
            const mainCommandFiles = await fs.readdir(path.join(__dirname, 'commands'));
            for (const file of mainCommandFiles) {
                if (file.endsWith('.js')) {
                    const command = require(path.join(__dirname, 'commands', file));
                    if (command.data && command.execute) {
                        this.commands.set(command.data.name, command);
                        console.log(`✅ Loaded command/${file}`);
                    }
                }
            }
        } catch (e) {
            console.error(`❌ Error loading main commands: ${e}`);
        }

        // Load subfolders
        const commandFolders = [
            'moderation',
            'roles',
            'utility',
            'help',
            'embed',
            'sticky'
        ];

        for (const folder of commandFolders) {
            const folderPath = path.join(__dirname, 'commands', folder);
            try {
                await fs.mkdir(folderPath, { recursive: true });
                const files = await fs.readdir(folderPath);
                for (const file of files) {
                    if (file.endsWith('.js')) {
                        const command = require(path.join(folderPath, file));
                        if (command.data && command.execute) {
                            this.commands.set(command.data.name, command);
                            console.log(`✅ Loaded ${folder}/${file}`);
                        }
                    }
                }
            } catch (e) {
                console.error(`❌ Error loading ${folder}: ${e}`);
            }
        }

        // Start auto-save
        this.dataStorage.autoSave();
    }

    async updateServerInfo(guild) {
        this.serverInfo.set(guild.id, {
            name: guild.name,
            id: guild.id,
            memberCount: guild.memberCount,
            iconUrl: guild.iconURL(),
            bannerUrl: guild.bannerURL(),
            description: guild.description || "No description set",
            owner: guild.owner,
            createdAt: guild.createdAt,
            roles: guild.roles.cache.size,
            channels: guild.channels.cache.size,
            boostLevel: guild.premiumTier,
            boostCount: guild.premiumSubscriptionCount
        });
    }

    async start() {
        await this.init();
        await this.login(TOKEN);
        
        this.user.setActivity('/help | Server Manager', { type: ActivityType.Playing });
        console.log('✅ Bot is online!');
    }
}

const bot = new Bot();

// Message handling
bot.on('messageCreate', async (message) => {
    if (message.author.bot) return;

    const channelId = message.channel.id;
    if (bot.stickyMessages.has(channelId)) {
        const now = Date.now();
        const lastSent = bot.stickyLastSent[channelId];
        const cooldown = bot.stickyCooldowns.get(channelId) || 1;

        if (lastSent && (now - lastSent) < 5) return; // 5ms cooldown

        const stickyData = bot.stickyMessages.get(channelId);

        // Delete previous sticky messages
        const messages = await message.channel.messages.fetch({ limit: 50 });
        for (const [, oldMessage] of messages) {
            if (oldMessage.author.id === bot.user.id &&
                (oldMessage.id === stickyData.messageId ||
                 (oldMessage.embeds.length > 0 && stickyData.isEmbed) ||
                 (oldMessage.embeds.length === 0 && !stickyData.isEmbed))) {
                try {
                    await oldMessage.delete();
                } catch (e) {
                    if (e.code !== 10008) console.error(e);
                }
            }
        }

        await new Promise(resolve => setTimeout(resolve, 100));

        let newSticky;
        const content = stickyData.content;

        if (typeof content === 'object' && content.type === 'stored_embed') {
            const embedData = content.original_data;
            const embed = new EmbedBuilder()
                .setTitle(embedData.title)
                .setDescription(embedData.description)
                .setColor(embedData.color);

            if (embedData.footer) {
                embed.setFooter({ text: `${embedData.footer} • 📌 Sticky message • ${message.guild.name}` });
            } else {
                embed.setFooter({ text: `📌 Sticky message • ${message.guild.name}` });
            }

            if (embedData.thumbnail) embed.setThumbnail(embedData.thumbnail);
            if (embedData.image) embed.setImage(embedData.image);

            newSticky = await message.channel.send({ embeds: [embed] });
        } else if (stickyData.isEmbed) {
            const embed = new EmbedBuilder()
                .setTitle(content.title || 'Sticky Message')
                .setDescription(content.description)
                .setColor(content.color || 'Blue')
                .setFooter({ text: `📌 Sticky message • ${message.guild.name}` })
                .setTimestamp();

            newSticky = await message.channel.send({ embeds: [embed] });
        } else {
            newSticky = await message.channel.send(stickyData.content);
        }

        bot.stickyLastSent[channelId] = now;
        await bot.updateStickyMessage(channelId, newSticky.id, stickyData.content);
    }
});

process.on('unhandledRejection', error => {
    console.error('Unhandled promise rejection:', error);
});

// Start the bot
(async () => {
    try {
        keepAlive();
        console.log("✅ Web server started!");
        console.log("✅ Starting bot with data persistence...");
        await bot.start();
    } catch (e) {
        console.error(`❌ Error starting bot: ${e}`);
        if (bot.dataStorage) {
            await bot.dataStorage.saveData();
        }
    }
})();
